fs/vfs/fs_timerfd.c

```diff
@@ -0,0 +1,824 @@
+      goto errout;
+    }
+
+  /* Notify the POLLIN event if the counter is not zero */
+
+  if (timerfd_get_counter(dev) > 0)
+    {
+      timerfd_pollnotify(dev, POLLIN);
+    }
+
+errout:
+  nxsem_post(&dev->exclsem);
+  return ret;
+}
+#endif
+
+static void timerfd_timeout_work(FAR void *arg)
+{
+  FAR struct timerfd_priv_s *dev = (FAR struct timerfd_priv_s *)arg;
+  int ret;
+
+  ret = nxsem_wait_uninterruptible(&dev->exclsem);
+  if (ret < 0)
+    {
+      wd_cancel(&dev->wdog);
+      return;
+    }
+
+#ifdef CONFIG_TIMER_FD_POLL
+  /* Notify all poll/select waiters */
+
+  timerfd_pollnotify(dev, POLLIN);
+#endif
+
+  /* Notify all of the waiting readers */
+
+  timerfd_waiter_sem_t *cur_sem = dev->rdsems;
+  while (cur_sem != NULL)
+    {
+      nxsem_post(&cur_sem->sem);
+      cur_sem = cur_sem->next;
+    }
+
+  dev->rdsems = NULL;
+
+  nxsem_post(&dev->exclsem);
+}
+
+static void timerfd_timeout(wdparm_t idev)
+{
+  FAR struct timerfd_priv_s *dev = (FAR struct timerfd_priv_s *)idev;
+  irqstate_t intflags;
+
+  /* Disable interrupts to ensure that expiration counter is accessed
+   * atomically
+   */
+
+  intflags = enter_critical_section();
+
+  /* Increment timer expiration counter */
+
+  dev->counter++;
+
+  work_queue(TIMER_FD_WORK, &dev->work, timerfd_timeout_work, dev, 0);
+
+  /* If this is a repetitive timer, then restart the watchdog */
+
+  if (dev->delay)
+    {
+      wd_start(&dev->wdog, dev->delay, timerfd_timeout, idev);
+    }
+
+  leave_critical_section(intflags);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+int timerfd_create(int clockid, int flags)
+{
+  FAR struct timerfd_priv_s *new_dev;
+  int new_fd;
+  int ret;
+
+  /* Sanity checks. */
+
+  if (clockid != CLOCK_REALTIME
+#ifdef CONFIG_CLOCK_MONOTONIC
+      && clockid != CLOCK_MONOTONIC
+      && clockid != CLOCK_BOOTTIME
+#endif /* CONFIG_CLOCK_MONOTONIC */
+      )
+    {
+      ret = EINVAL;
+      goto errout;
+    }
+
+  /* devpath: TIMER_FD_VFS_PATH + /tfd (4) + %u (10) + null char (1) */
+
+  char devpath[sizeof(CONFIG_TIMER_FD_VFS_PATH) + 4 + 10 + 1];
+
+  /* Allocate instance data for this driver */
+
+  new_dev = timerfd_allocdev();
+  if (new_dev == NULL)
+    {
+      /* Failed to allocate new device */
+
+      ret = ENOMEM;
+      goto errout;
+    }
+
+  /* Initialize the timer instance */
+
+  new_dev->clock = clockid;
+  new_dev->crefs = 1;
+  new_dev->delay = 0;
+  new_dev->counter = 0;
+
+  /* Request a unique minor device number */
+
+  new_dev->minor = timerfd_get_unique_minor();
+
+  /* Get device path */
+
+  sprintf(devpath, CONFIG_TIMER_FD_VFS_PATH "/tfd%u", new_dev->minor);
+
+  /* Register the driver */
+
+  ret = register_driver(devpath, &g_timerfd_fops, 0444, new_dev);
+  if (ret < 0)
+    {
+      ferr("Failed to register new device %s: %d\n", devpath, ret);
+      ret = -ret;
```

Comment:
remove, let's assign ret to -EINVAL and -ENOMEM and set_errno(-ret) instead

Reply:
What do we gain with this?

Reply:
- `int closedir(FAR DIR *dirp)`
- `FAR DIR *opendir(FAR const char *path)`
- `FAR struct dirent *readdir(DIR *dirp)`
- `FAR char *getenv(FAR const char *name)`
- and others
Use the same approach as here

Reply:
I do not mind, but probably will need to update those as well in next PR

Reply:
> What do we gain with this?

Could save ret = -ret in several places

Reply:
Ok. I will redo the logic

Reply:
Done